penty/pentypes/builtins.py

```diff
@@ -529,13 +532,205 @@ def dict_instanciate(ty):
 ##
 #
 
+def check_self_ty(expected_ty):
```

Comment:
That's a great idea. Probably something we could move in types.py, providing a new `MethodType` and that type would ensure the check happens.